File: src/initialSubsets.cpp
/*
 * Author: Andreas Alfons
 *         Erasmus University Rotterdam
 */

#include "initialSubsets.h"

using namespace Rcpp;
using namespace arma;

// function used internally, which computes lasso fits for subsets containing a 
// small number of observations (typically only 3) and returns the indices of 
// the respective h observations with the smallest absolute residuals
umat sparseSubsets(const mat& x, const vec& y, const double& lambda, 
    const uword& h, const umat& subsets, const bool& normalize, 
    const bool& useIntercept, const double& eps, const bool& useGram) {
	const uword nsamp = subsets.n_cols;
	umat indices(h, nsamp);
	for(uword k = 0; k < nsamp; k++) {
		// compute lasso fit
		double intercept;
		vec coefficients = fastLasso(x, y, lambda, true, subsets.unsafe_col(k),
				normalize, useIntercept, eps, useGram, intercept);
		// compute residuals
		vec residuals = y - x * coefficients;
		if(useIntercept) {
			residuals -= intercept;
		}
		// find h observations with smallest absolute residuals
		indices.col(k) = findSmallest(abs(residuals), h);
	}
	return indices;
}

// R interface to sparseSubsets()
SEXP R_sparseSubsets(SEXP R_x, SEXP R_y, SEXP R_lambda, SEXP R_h, 
    SEXP R_subsets, SEXP R_normalize, SEXP R_intercept, SEXP R_eps, 
    SEXP R_useGram) {
	// data initializations
	NumericMatrix Rcpp_x(R_x);						  // predictor matrix
	const int n = Rcpp_x.nrow(), p = Rcpp_x.ncol();
	mat x(Rcpp_x.begin(), n, p, false);     // reuse memory
	NumericVector Rcpp_y(R_y);			        // response
	vec y(Rcpp_y.begin(), n, false);        // reuse memory
  double lambda = as<double>(R_lambda);
  int h = as<int>(R_h);
	IntegerMatrix Rcpp_subsets(R_subsets);  // subset to use for computation
	const int s = Rcpp_subsets.nrow(), nsamp = Rcpp_subsets.ncol();
	umat subsets(s, nsamp);
	for(int j = 0; j < nsamp; j++) {
		for(int i = 0; i < s; i++) {
      // can't use the same memory-saving conversion for integer matrices
			subsets(i,j) = Rcpp_subsets(i,j) - 1;
		}
	}
  bool normalize = as<bool>(R_normalize);
  bool useIntercept = as<bool>(R_intercept);
	double eps = as<double>(R_eps);
	bool useGram = as<bool>(R_useGram);
	// call native C++ function and return results
	umat indices = sparseSubsets(x, y, lambda, h, subsets, 
      normalize, useIntercept, eps, useGram) + 1;
	return wrap(indices);
}

File: R/initialSubsets.R
# ------------------------------------
# Author: Andreas Alfons
#         Erasmus University Rotterdam
# ------------------------------------

## generate random subsets
randomSubsets <- function(n, h, nsamp = 500) replicate(nsamp, sample.int(n, h))

## generate random subsets based on hyperplane
hyperplaneSubsets <- function(x, y, h, nsamp = 500) {
  d <- dim(x)
  x <- addIntercept(x)
  subsets <- randomSubsets(d[1], d[2], nsamp)
  subsets <- apply(subsets, 2, function(i, x) {
    xi <- x[i, , drop=FALSE]  # select observations
    theta <- try(hyperplane(xi), silent=TRUE)  # hyperplane coefficients or error message
    # check if point form hyperplane and add points if not
    while(inherits(theta, "try-error") && length(i) < d[1]) {
      ind <- (seq_len(d[1]))[-i]  # not yet sampled observations
      new <- if(length(ind) > 1) sample(ind, 1) else ind
      i <- c(i, new)
      xi <- rbind(xi, x[new,])  # add new point
      theta <- try(hyperplane(xi), silent=TRUE)  # hyperplane coefficients or error message
    }
    if(inherits(theta, "try-error")) {
      # singularity issues even with all data points
      sample.int(d[1], h)  # random subset
    } else {
      # compute residuals w.r.t. hyperplane
      residuals <- y - x %*% theta
      # find observations with smallest absolute residuals
      findSmallest(abs(residuals), h)
    }
  }, x)
}

## generate subsets based on lasso solutions with 3 observations
sparseSubsets <- function(x, y, lambda, h, nsamp = 500, intercept = TRUE, 
                          eps = .Machine$double.eps, use.Gram = TRUE) {
  # obtain random subsets with only 3 observations
  n <- length(y)
  subsets <- randomSubsets(n, 3, nsamp)
  # call C++ function to compute lasso fits and find observations with 
  # smallest absolute residuals
  callBackend <- getBackend()
  subsets <- callBackend("R_sparseSubsets", R_x=x, R_y=y, R_lambda=lambda, 
                         R_h=h, R_subsets=subsets, R_intercept=intercept, 
                         R_eps=eps, R_useGram=use.Gram)
}

# ## generate initial subsets
# initialSubsets <- function(x, y, h, nsamp = 500, 
#                            type = c("sparse", "hyperplane", "random"),
#                            lambda, intercept = TRUE, eps = .Machine$double.eps, 
#                            use.Gram = TRUE) {
#   ## initializations
#   d <- dim(x)
#   type <- match.arg(type)
#   if(h < d[2] && type == "hyperplane") type <- "sparse"
#   ## generate subsets
#   if(type == "sparse") {
#     # obtain random subsets with only 3 observations
#     subsets <- randomSubsets(d[1], 3, nsamp)-1
#     # call C++ function to compute lasso fits and find observations with 
#     # smallest absolute residuals
#     callBackend <- getBackend()
#     subsets <- callBackend("R_initialSubsetsSparse", R_x=x, R_y=y, 
#                            R_subsets=subsets, R_h=h, R_lambda=lambda, R_intercept=intercept, 
#                            R_eps=eps, R_useGram=use.Gram)+1
#   } else if(type == "hyperplane") {
#     ## subsets based on hyperplane
#     x <- addIntercept(x)
#     subsets <- randomSubsets(d[1], d[2], nsamp)
#     subsets <- apply(subsets, 2, function(i, x) {
#       xi <- x[i, , drop=FALSE]  # select observations
#       theta <- try(hyperplane(xi), silent=TRUE)  # hyperplane coefficients or error message
#       # check if point form hyperplane and add points if not
#       while(inherits(theta, "try-error") && length(i) < d[1]) {
#         ind <- (seq_len(d[1]))[-i]  # not yet sampled observations
#         new <- if(length(ind) > 1) sample(ind, 1) else ind
#         i <- c(i, new)
#         xi <- rbind(xi, x[new,])  # add new point
#         theta <- try(hyperplane(xi), silent=TRUE)  # hyperplane coefficients or error message
#       }
#       if(inherits(theta, "try-error")) {
#         # singularity issues even with all data points
#         sample.int(d[1], h)  # random subset
#       } else {
#         # compute residuals w.r.t. hyperplane
#         residuals <- y - x %*% theta
#         # find observations with smallest absolute residuals
#         findSmallest(abs(residuals), h)
#       }
#     }, x)
#   } else {
#     ## random subsets
#     subsets <- randomSubsets(d[1], h, nsamp)
#   }
#   ## return subsets
#   subsets
# }
